add permute frontends.torch

PatternFinder opened the pull request:
link to issue: #2792

Daniel4078 reviewed (changes requested):
Hello, I found a small error in the test you wrote, here is the detail: ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:91:17: F821 undefined name 'ivy_np'
Please fix that, Thank you.

PatternFinder commented:
> Hello, I found a small error in the test you wrote, here is the detail: ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:91:17: F821 undefined name 'ivy_np' Please fix that, Thank you.

ok, forgot to add
`import ivy.functional.backends.numpy as ivy_np`
now added, ready for more review. thanks

Daniel4078 commented:
just a small tip, you can re-request a fresh review from a reviewer by click the refresh icon in the sidebar of the Conversation tab.

PatternFinder commented:
> just a small tip, you can request a fresh review from a reviewer by click the refresh icon in the Reviewers section of the sidebar of the Conversation tab. Now I will wait till all the checks on the pull request are done and then decide if there is any more problem.

  thanks for the tip! appreciate any feedbacks 👍

Daniel4078 reviewed (changes requested):
According to the checks, the test you write have these errors:
1 TypeError: test_frontend_function() missing 1 required keyword-only argument: 'with_out'
2 you did not use helpers.dtype_and_values correctly
Thank you!

PatternFinder commented:
> According to the checks, the test you write have these errors: 1 TypeError: test_frontend_function() missing 1 required keyword-only argument: 'with_out' 2 you did not use helpers.dtype_and_values correctly Thank you!

ok I added the with_out argument and switched to helpers.dtype_values_axis for the axis input; please have a look thanks

Daniel4078 commented on ivy/functional/frontends/torch/indexing_slicing_joining_mutating_ops.py:
```diff
@@ -4,3 +4,14 @@
+
+
+def permute(input, dims):
+    return ivy.permute_dims(input, dims)
```
I just confirmed that ivy.permute_dims have a out argument just like ivy.concat(), so you need to add out=Non argument to your permute function and put that out in to the ivy.permute_dims() just like the cat function right before your code.

Daniel4078 commented on ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:
```diff
@@ -70,3 +71,61 @@ def test_torch_cat(
+        fw,
+):
+    dtype, value, axis = dtype_values_axis
+    helpers.test_frontend_function(
```
please read the description of helpers.test_frontend_function here:
"""Tests a frontend function for the current backend by comparing the result with
    the function in the associated framework.

    Parameters
    ----------
    input_dtypes
        data types of the input arguments in order.
    as_variable_flags
        dictates whether the corresponding input argument should be treated
        as an ivy Variable.
    with_out
        if True, the function is also tested with the optional out argument.
    num_positional_args
        number of input arguments that must be passed as positional
        arguments.
    native_array_flags
        dictates whether the corresponding input argument should be treated
        as a native array.
    fw
        current backend (framework).
    frontend
        current frontend (framework).
    fn_name
        name of the function to test.
    rtol
        relative tolerance value.
    atol
        absolute tolerance value.
    test_values
        if True, test for the correctness of the resulting values.
    all_as_kwargs_np
        input arguments to the function as keyword arguments.

    Returns
    -------
    ret
        optional, return value from the function
    ret_np
        optional, return value from the Numpy function
    """
You should not randomly generate the as_variable_flags, with_out, native_array_flags, container_flags and instance_method arguments as they should be determined according to the situation like this example here:
@given(
    dtype_and_x=helpers.dtype_and_values(available_dtypes=ivy_np.valid_float_dtypes),
    num_positional_args=helpers.num_positional_args(fn_name="variable"),
    data=st.data(),
)
@handle_cmd_line_args
def test_variable(
    *,
    dtype_and_x,
    num_positional_args,
    native_array,
    container,
    instance_method,
    fw,
):
    dtype, x = dtype_and_x
    helpers.test_function(
        input_dtypes=dtype,
        as_variable_flags=True,
        with_out=False,
        num_positional_args=num_positional_args,
        native_array_flags=native_array,
        container_flags=container,
        instance_method=instance_method,
        fw=fw,
        fn_name="variable",
        x=np.asarray(x, dtype=dtype),
    )

PatternFinder replied:
I see, how would you set them? would these be correct?
`as_variable=helpers.array_bools()`
`with_out`: for function that has this optional out, what should I set it to? 
`native_array=helpers.array_bools(),` 
how should I set these in this case? `container_flags` and `instance_method`

Daniel4078 replied:
In this case, the permute function does have an out argument, so the with_out argument should always be True.
Finally, as_variable_flags, native_array_flags should be set automatically, by adding @handle_cmd_line_args before the test function and have arguments in your test function receiving the values and passing into test_frontend_function.

Daniel4078 reviewed (changes requested):
Please deal with the two comments I have with your code in the Files changed section of the pull request.

Daniel4078 reviewed (changes requested):
Maybe the example I gave you was not so accurate, so please follow the comment I gave. If you are not sure what to do, try to find examples on your own by looking at other tests done by other people. If you have any question, please ask, thank you.

Daniel4078 commented:
By the way, you can ask questions about how to write this on our discord channel as well.

PatternFinder commented:
> By the way, you can ask questions about how to write this on our discord channel as well.

I've tried to ask questions in the beginning, but even in the #general channal, I didn't get any responses, is it because I've always just @ivyteam? who would be a better person to @ in discord for questions?

PatternFinder commented:
if any function includes a `out` argument, wouldn't we want to generate both True and False to test it? why would it need to always be set to True?

PatternFinder commented:
added the out arguments and adjusted the test function, ready for another look

Daniel4078 commented on ivy/functional/frontends/torch/indexing_slicing_joining_mutating_ops.py:
```diff
@@ -4,3 +4,14 @@
     return ivy.concat(tensors, dim, out=out)
+
+
+def permute(input, dims, *, out=None):
```
Sorry for my mistake before, I just found that torch.permute doesn't accept an out argument, so there need not be one in the tests

Daniel4078 commented on ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:
```diff
@@ -70,3 +72,42 @@ def test_torch_cat(
+
+# permute
+@given(
+    dtype_values_axis=helpers.dtype_values_axis(
```
Also, dims is supposed to be a tuple of ints, so the data generated for axis should be a tuple instead of an axis(integer)

PatternFinder replied:
ok I added a ret_tuples to the helpers.get_axis function; this should only return a tuple now when turned on;

Daniel4078 reviewed (changes requested):
I am sorry for a mistake I made before as I found that torch.permute does not have an out argument. After you fix these two places, I suppose there will be not more errors. Thank you.

Daniel4078 reviewed (approved):
I suppose it is good enough now. Thank you.

Daniel4078 reviewed (changes requested):
I just found that there are some formatting issues needs to be fixed before merging. Thank you.
ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:[9](https://github.com/unifyai/ivy/runs/7818549741?check_suite_focus=true#step:5:10)2:5: E122 continuation line missing indentation or outdented
ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:92:9: E251 unexpected spaces around keyword / parameter equals
ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:92:11: E251 unexpected spaces around keyword / parameter equals
ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:92:89: E501 line too long ([10](https://github.com/unifyai/ivy/runs/7818549741?check_suite_focus=true#step:5:11)9 > 88 characters)

PatternFinder commented:
fixed outdents

Daniel4078 reviewed (approved):
I have fixed the last formatting and docstring errors and it is good enough to me. Thank you.

Daniel4078 commented on ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:
```diff
@@ -70,3 +72,41 @@ def test_torch_cat(
+            set(ivy_np.valid_float_dtypes).intersection(
+                set(ivy_torch.valid_float_dtypes))),
+        shape=st.shared(helpers.get_shape(min_num_dims=1), key="shape"),
+    axis = helpers.get_axis(shape=st.shared(helpers.get_shape(min_num_dims=1), key="shape"), ret_tuple=True),
```
it seems that there is one extra "," at the end of this line, is this an error?

Daniel4078 replied:
Also this line have some formatting issues:
ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:92:5: E122 continuation line missing indentation or outdented
ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:92:9: E251 unexpected spaces around keyword / parameter equals
ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:92:11: E251 unexpected spaces around keyword / parameter equals
ivy_tests/test_ivy/test_frontends/test_torch/test_indexing_slicing_joining_mutating_ops.py:92:89: E501 line too long (109 > 88 characters)

PatternFinder replied:
> it seems that there is one extra "," at the end of this line, is this an error?

```
    axis = helpers.get_axis(
        shape=st.shared(helpers.get_shape(min_num_dims=1), key="shape"), 
        ret_tuple=True),
```
you mean the `,` after `ret_tuple`? `ret_tuple` is the arg I added to the helpers.get_axis function

Daniel4078 reviewed (changes requested):
Please fix that one line I pointed out in the comment and there are no other problems. Thanks.

Daniel4078 commented:
# permute
@given(
    dtype_values_axis=helpers.dtype_values_axis(
        available_dtypes=tuple(
            set(ivy_np.valid_float_dtypes).intersection(
                set(ivy_torch.valid_float_dtypes))),
        shape=st.shared(helpers.get_shape(min_num_dims=1), key="shape"),
        axis=helpers.get_axis(
            shape=st.shared(helpers.get_shape(min_num_dims=1), key="shape"),
            ret_tuple=True),
    ),
    as_variable=helpers.array_bools(),
    num_positional_args=helpers.num_positional_args(
        fn_name="ivy.functional.frontends.torch.permute"),
    native_array=helpers.array_bools(),
)
def test_permute(
        dtype_values_axis,
        as_variable,
        num_positional_args,
        native_array,
        fw,
):
    dtype, value, axis = dtype_values_axis
    helpers.test_frontend_function(
        input_dtypes=dtype,
        as_variable_flags=as_variable,
        with_out=False,
        num_positional_args=num_positional_args,
        native_array_flags=native_array,
        fw=fw,
        frontend="torch",
        fn_name="permute",
        input=np.asarray(value, dtype=dtype),
        dims=axis,
    )

Daniel4078 commented:
I just used pycharm to auto format your code, please have a look.

Daniel4078 commented:
I suppose that the line number given by the github checks are wrong

PatternFinder commented:
ok I figured out of the E122 problem. I missed a ')' on helpers.dtype_values_axis, causing helpers.get_axis to be included as one of helpers.dtype_values_axis's arguments, it should be fixed now.

I see that lint is passed

Daniel4078 commented:
I can fix the formatting for you if you want

Daniel4078 reviewed (approved):
Now I suppose it is good to be merged now as the other failed checks are not caused by your changes.

PatternFinder commented:
> I can fix the formatting for you if you want

I think the given() part's formatting should be correct now. `helpers.dtype_values_axis` and `helpers.get_axis` are now correct two separately given
